torch/csrc/jit/tensorexpr/bounds_overlap.cpp

```diff
@@ -14,16 +49,43 @@ OverlapKind boundOverlap(Bound a, Bound b) {
     return ContainedOrEqual;
   }
 
+  // We have to figure out if the bounds fall under the following 2 cases:
+  // 1. a is before b
+  //      a.start ... a.end ... b.start ... b.end
+  // 2. b is before a
+  //      b.start ... b.end ... a.start ... a.end
+  //
+  // So, we compute "a.start - b.end" and "b.start - a.end". If even one of
+  // those is positive, then it is guaranteed that the bounds do not overlap.
+  //
+  // If the diff is a constant, then we can directly check if the constant is
+  // positive. If the diff is not a constant, then it will be made of
+  // variables that correspond to the bounds of buffers involved. These buffer
+  // bounds can never be negative. So, we check if the given expression is
+  // guaranteed to be positive under the assumption that the variables involved
+  // are never negative.
+
   ExprPtr lowDiff = IRSimplifier::simplify(alloc<Sub>(a.start, b.end));
   ExprPtr highDiff = IRSimplifier::simplify(alloc<Sub>(b.start, a.end));
 
-  if (lowDiff->isConstant() && highDiff->isConstant()) {
+  if (lowDiff->isConstant()) {
     int low = immediateAs<int>(lowDiff);
+    // No overlap.
+    if (low > 0) {
```

Comment:
This condition only works when the "gap" is more than 0, while the condition below seem to be true for 0. If that's so, can we make it consistent? Also, do we even need this if? Can it all be handled by a single `isPositiveExpr` check?

Reply:
Good point. I didn't think about the difference earlier. Thanks for pointing that out.

We need to ensure that it is positive because 0 means there is some overlap. I have modified the code to look for 0 as well now and identify only the guaranteed positive cases. PTAL.

Reply:
When we have `a.start -- a.end` do we actually include `a.end`? Not allowing 0 seems to result in a hairy code now, so maybe we should just make our ranges to not include the last value.

Reply:
Unfortunately, our bounds data structure is set up to include `a.end`. I'm afraid changing that would involve touching a lot of the mem analysis code base. 

In the current setup, we should not be allowing 0. Here is the approximation that I currently have:
 
The given expression is guaranteed to be positive if all these conditions hold:
   * does not include any "subtract" term
   * does not include any "div" term
   * all immediate values are positive (> 0)

I see that this is not quite enough since an expression can get to 0 with mod, shift, xor, etc. 

I'm thinking that we should be even more conservative and allow the expressions that only contain:
  * "add" and "mul" terms
  * all immediate values are positive.
Though, implementing this is going to involve checking all the cases that should not be allowed, which is frustrating.

What do you think?

Reply:
I wasn't able to convince myself that this handles all the corner cases. Also, implementing this (checking for all such possible cases that could lead to a zero or negative value) is cumbersome and errorprone. So, I'm removing the checks with variables in this.